src/OfflineTool.cpp

```diff
@@ -0,0 +1,473 @@
+
+    std::cout << result << std::endl;
+}
+
+void
+doDeserialize(std::string const& data)
+{
+    auto const fail = [&]
+    {
+        std::cout << "Unable to deserialize \"" << data << "\"" <<
+            std::endl;
+    };
+    try
+    {
+        auto const result = serialize::deserialize(boost::trim_copy(data));
+
+        if (result)
+            std::cout << result->getJson(0).toStyledString() << std::endl;
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
+    }
+}
+
+void
+doSingleSign(std::string const& data,
+    boost::filesystem::path const& keyFile)
+{
+    using namespace serialize;
+    auto const fail = [&]()
+    {
+        std::cout << "Unable to sign \"" << data << "\"" << std::endl;
+    };
+    try
+    {
+        auto tx = make_sttx(boost::trim_copy(data));
+        if (tx)
+        {
+            auto const rippleKey = RippleKey::make_RippleKey(keyFile);
+
+            rippleKey.singleSign(*tx);
+
+            std::cout << tx->getJson(0).toStyledString() << std::endl;
+        }
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
+    }
+}
+
+void
+doMultiSign(std::string const& data,
+    boost::filesystem::path const& keyFile)
+{
+    using namespace serialize;
+    auto const fail = [&]()
+    {
+        std::cout << "Unable to sign \"" << data << "\"" << std::endl;
+    };
+    try
+    {
+        auto tx = make_sttx(boost::trim_copy(data));
+        if (tx)
+        {
+            auto rippleKey = RippleKey::make_RippleKey(keyFile);
+
+            rippleKey.multiSign(*tx);
+
+            std::cout << tx->getJson(0).toStyledString() << std::endl;
+        }
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
+    }
+}
+
+void
+doCreateKeyfile(boost::filesystem::path const& keyFile,
+    boost::optional<std::string> const& keytype,
+    boost::optional<std::string> const& seed)
+{
+    using namespace ripple;
+    using namespace serialize;
+
+    if (exists(keyFile))
+        throw std::runtime_error(
+            "Refusing to overwrite existing key file: " +
+            keyFile.string());
+
+    auto const kt = keytype ?
+        boost::optional<KeyType>{keyTypeFromString(*keytype)} :
+        boost::none;
+    if (kt && *kt == KeyType::invalid)
+    {
+        std::cout << "Invalid key type: \"" << *keytype << "\"" <<
+            std::endl;
+        return;
+    }
+
+    auto key = RippleKey::make_RippleKey(kt, seed);
+
+    key.writeToFile(keyFile);
+
+    std::cout << "New ripple key created " <<
+        "in " << keyFile.string() << "\n" <<
+        "Key type is " << to_string(key.keyType()) << ", and " <<
+        "account ID is " <<
+        toBase58(calcAccountID(key.publicKey())) << "\n" <<
+        "\nThis file should be stored securely and not shared\n\n";
+
+}
+
+std::string
+getStdin()
+{
+    std::ostringstream stdinput;
```

Comment:
Another thought on the apparently hung command if not enough parameters are passed.  A much simpler solution that throwing a thread at the problem would be to add a `--stdin` optional argument.  If the user specifies `--stdin` then the argument must come from `stdin`.  If they don't specify `--stdin`, then the argument must be provided on the command line (so an error message can be provided if its missing).  If they intentionally add the `--stdin` flag then presumably they understand about standard in.

Waddya think?